crates/symbolicator-service/src/download/index.rs

```diff
@@ -0,0 +1,419 @@
+/// which files are available on that source. If we have an index available,
+/// we don't even need to make requests to the source for files not in
+/// the index.
+///
+/// This service takes care of fetching indexes from sources which indicate
+/// that they provide one (via the `has_index` field). The type of index is
+/// internally determined by the source's layout, although only Symstore is
+/// supported for now.
+#[derive(Debug, Clone)]
+pub struct SourceIndexService {
+    /// The cache for storing Symstore indexes.
+    symstore_cache: Arc<Cacher<FetchSymstoreIndex>>,
+    /// The cache used for storing individual Symstore index segments.
+    symstore_segment_cache: Arc<Cacher<FetchSymstoreIndexSegment>>,
+    /// An in-memory cache for keeping track of the last segment uploaded
+    /// to a Symstore index.
+    symstore_last_id_cache: moka::future::Cache<(Scope, SourceId), CacheContents<u32>>,
+    /// The download service to download index files.
+    downloader: Arc<DownloadService>,
+}
+
+impl SourceIndexService {
+    /// Creates a new `SourceIndexService`.
+    ///
+    /// This service will use the same cache for computing
+    /// both symstore index segments and entire symstore indexes.
+    pub fn new(
+        cache: Cache,
+        shared_cache: SharedCacheRef,
+        downloader: Arc<DownloadService>,
+    ) -> Self {
+        // Create an in-memory cache for Symstore last IDs.
+        // This is so we don't ask for the last ID on every request.
+        let last_id_cache = moka::future::Cache::builder()
+            .max_capacity(100)
+            .name("last_id")
+            .time_to_live(Duration::from_secs(24 * 60 * 60))
+            .build();
+
+        Self {
+            symstore_cache: Arc::new(Cacher::new(cache.clone(), shared_cache.clone())),
+            symstore_segment_cache: Arc::new(Cacher::new(cache, shared_cache)),
+            symstore_last_id_cache: last_id_cache,
+            downloader,
+        }
+    }
+
+    /// Fetches the ID of the most recently uploaded Symstore index
+    /// segment.
+    ///
+    /// This ID is stored in a file called `lastid.txt` in the
+    /// `000Admin` directory.
+    ///
+    /// The last ID is locally cached for an hour, so if a new
+    /// file is uploaded, we will see it at most an hour later.
+    async fn fetch_symstore_last_id(
+        &self,
+        scope: Scope,
+        source: &IndexSourceConfig,
+    ) -> CacheContents<u32> {
+        self.symstore_last_id_cache
+            .get_with((scope, source.id().clone()), async {
+                let mut temp_file = NamedTempFile::new()?;
+                let remote_file = source.remote_file(SourceLocation::new(LASTID_FILE));
+                self.downloader
+                    .download(remote_file, temp_file.path().to_path_buf())
+                    .await?;
+
+                let mut buf = Vec::new();
+                temp_file.read_to_end(&mut buf)?;
```

Comment:
That seems risky, possibly loading a multiple gigabyte file into memory.

Reply:
Fair enough, I wasn't considering adversarial input. Shoud be replaced with an mmap easily enough.